src/HexagonOptimize.cpp

```diff
@@ -1118,6 +1201,221 @@ class OptimizePatterns : public IRMutator {
+                if (!is_stencil_reduction(a, rfac)) {
+                    continue;
+                }
+                // Split a into a0, a1 to get the correct vector args
+                // for sliding window reduction instructions. Below are
+                // required shuffle indices for a0 and a1:
+                // For rfac == 2:
+                // If a  -> shuff[0, 1,...., out_lanes]
+                //    a0 -> shuff[0, 1,...., out_lanes - 1]
+                //    a1 -> shuff[2, 3,...., out_lanes + 1]
+                //          Last index of a1 is don't care
+                // For rfac == 3:
+                // If a  -> shuff[0, 1,...., out_lanes + 1]
+                //    a0 -> shuff[0, 1,...., out_lanes - 1]
+                //    a1 -> shuff[2, 3,...., out_lanes + 1]
+                // For rfac == 4:
+                // If a  -> shuff[0, 1,...., out_lanes + 3]
+                //    a0 -> shuff[0, 1,...., out_lanes - 1]
+                //    a1 -> shuff[4, 5,...., out_lanes + 4]
+                //          Last index of a1 is don't care
+                if (const Shuffle *shuff = a.as<Shuffle>()) {
+                    vector<int> a0_indices(out_lanes), a1_indices(out_lanes);
+                    for (int i = 0; i < out_lanes; i++) {
+                        a0_indices[i] = shuff->indices[i * rfac];
+                        a1_indices[i] = shuff->indices[(i + 1) * rfac - 1];
+                    }
+                    a0 = Shuffle::make(shuff->vectors, a0_indices);
+                    a1 = Shuffle::make(shuff->vectors, a1_indices);
+                    if (rfac == 2 || rfac == 4) {
+                        // We'll need to rotate the indices by one element
+                        // to get the correct order.
+                        Type ty = UInt(8).with_lanes(a1.type().lanes() * a1.type().bytes());
+                        a1 = reinterpret(a1.type(),
+                                         Call::make(ty, "halide.hexagon.vror",
+                                                    {reinterpret(ty, a1), a1.type().bytes()},
+                                                    Call::PureExtern));
+                    } else {
+                        // Vtmpy has additional requirement that third
+                        // coefficient b[2] needs to be 1.
+                        if (!can_prove(Shuffle::make_extract_element(b, 2) == 1)) {
+                            continue;
+                        }
+                        b = Shuffle::make_slice(b, 0, 1, 2);
+                        b_ty = b.type();
+                    }
+                    a = Shuffle::make_concat({a0, a1});
+                } else {
+                    continue;
+                }
+            }
+            // Reinterpret scalar b arg to get correct type.
+            if (sig.flags & Signature::ScalarB) {
+                b = simplify(reinterpret(Type(b_ty.code(), b_ty.lanes() * b_ty.bits(), 1), b));
```

Comment:
Is this making a 24-bit int sometimes, e.g. for vtmpy? I guess this is OK, but this wasn't supported until very recently.